### ISSUE TITLE
Request for SFPU LLKs for more flexible broadcasting

### ISSUE THREAD

**Issue by yan-zaretskiy** [tenstorrent/tt-metal#16103]
Currently we have 3 broadcasting LLKs (e.g. `add_tiles_bcast`) that support subtile broadcasting of the second operand. This is not sufficient for a general binary op, because we need to support broadcasting cases like `A(1, 32) + B(32, 1)`. Here `A` contains a single row and `B` a single column. To enable this, I had to implement broadcasting in the data movement kernels: 

https://github.com/tenstorrent/tt-metal/blob/main/ttnn/cpp/ttnn/operations/eltwise/binary_ng/device/kernels/dataflow/fill_tile_utils.hpp

This is not ideal for obvious reasons: 
1. Hurts performance 
2. **Does not work** for block data types. 

Therefore, I would like to request a new LLK that would replicate the row/col/elements of a tile to the entire tile using SFPU. @rtawfik01 @rdjogoTT

**Comment by amahmudTT** [comment]
@yan-zaretskiy We need to assess the priority of this issue. Two questions. 

1. Is it blocking any op or in other words how much of a performance/functionality bottleneck is this issue. 

2. If we were to implement it were you expecting something like 

     add_tiles_AB_bcast <bcastType btypeA , bcastType btypeB> (in_cb_A, in_cb_B, out_cb) 

 where the bcastType can be either column or row, and we then have the ability to mix and match broadcast type for A and B, but we assume some form of broadcast will be there.  Or you had some other way in mind ?

**Comment by amahmudTT** [comment]
@yan-zaretskiy Do you want the flexibility to broadcast for all binary ops, or you have a specific set of ops for which you need this. Asking because the data movement pattern may be different for different binary ops and the ideal data movement pattern in the tenxis enging for broadcasting for different ops may need to be different.

**Comment by amahmudTT** [comment]
Notes from talk with Yan and Patrick. FYI @ttmtrajkovic 

For the ops that require the above mentioned feature. Add Mul Sub ops will be enough. 

The critical and most important reason we need this feature feature is that it is **blocking the use of block formats,** since they cannot be broadcasted by the current workarounds. 

But for other ops apart from Add, Mul, Sub and not necessarily binary ops only,  if it is too complicated to implement them with the generalized broadcasting schemes, a unary broadcast op would be helpful and acceptable. So that the user can broadcast a single tile and use it in other ops. This provides maximum flexibility. But for performance reasons a single function allowing generalized broadcast for Add, Mul, Sub is still desired, since it will be possible to avoid some writes into L1 memory. 

Yan will try to provide us with rough estimates on how much performance he is sacrificing by using the workaround schemes.

**Comment by amahmudTT** [comment]
I am working on this issue. I plan to implement the unary broadcast first and then the flexible versions for the Add, Mul and Sub ops.

**Comment by eyonland** [comment]
@amahmudTT , any updates on this?

**Comment by amahmudTT** [comment]
> [@amahmudTT](https://github.com/amahmudTT) , any updates on this?

Working on the unary broadcast. ETA is end of next week. Then move on to the generalized version of add/mul/sub

**Comment by amahmudTT** [comment]
Yeah the unary one is now working for passthrough/col/row and scalar broadcast. I tested on wormhole, will test on blackhole and make the PR available before Monday.

**Comment by amahmudTT** [comment]
@yan-zaretskiy @patrickroberts 

Please have a look at this PR https://github.com/tenstorrent/tt-metal/pull/17479/files , this is for unary broadcast which you should be able to use in the general case. For the binary case I think this along with some existing functions (will require modification for multiplication, but hopefully will be okay for ADD and SUB) will do the job more efficiently than having a new api to do the binary operation on one pair of tiles at a time. Will sync with any one or both of you sometime regarding that. Thanks.

**Comment by amahmudTT** [comment]
Had a discussion with @patrickroberts 

After unary broadcasting that tile may need to be used multiple times. But for best performance in binary operations with broadcast, the ideal path would be to do the unary broadcast on multiple tiles and not pack the tiles out of dest and then reconfigure unpacker and packer and then do the multiple math ops which will overwrite the data in dest. So given we are likely to pack the data anyways from dest register it might be better to use the unary broadcast then pack the data out and then use it along with existing infrastructure that can broadcast the second operand. Patrick will let me know how the unary one performs in their use cases and we can decide whether to implement the binary one or not since if we pack the data out anyways we wont be saving much on performance with a specialized binary one.

**Comment by amahmudTT** [comment]
Since Patrick and Yan mentioned they are specifically interested in bfp formats, I started making my test case flexible to test many combinations, so far the testing is showing a somewhat weird behavior we saw in another issue. When I send fp16 data but broadcast and pack out bfp data the data is rounded using ceiling. This is has been observed in other bugs too for example 

https://github.com/tenstorrent/tt-metal/issues/17018

**Comment by amahmudTT** [comment]
The above rounding issue was found to be not a bug since the host side was modified to match pytorch's rounding to even and diverged from the device. So I will keep using this tolerance for this range.

**Comment by amahmudTT** [comment]
Since the use of bfp was stressed on I added test cases to extensivey test bfp8 format, and fp16b format, switching between them and unpacking as one and packing as nother and reconfiguring between them. They all pass given we use the tolerance based on the comment above.

**Comment by amahmudTT** [comment]
Just realized the broadcast wont work with fp32 without loss of precision, that is we need the unpacker dst format to be something other than fp32, so for fp32 where you need the precision to be preserved please use your current implementation.

**Comment by amahmudTT** [comment]
@patrickroberts I merged the unary part to main, if you find that part is enough for your tasks please close this issue. FYI FP32 without precision loss will not work as the hardware wont be able to handle it.

**Comment by vmilicevicTT** [comment]
@patrickroberts , could you please review last comments from @amahmudTT  and provide your feedback
FYI @cmaryanTT

**Comment by patrickroberts** [comment]
@amahmudTT Thanks for your work on this. Last time we discussed this I thought we concluded that the unary broadcasting would be sufficient for our use-case. I still plan to deliver performance comparisons of before and after, but not until I've finished resolving the remaining model issues on https://github.com/tenstorrent/tt-metal/pull/17160. I'm closing this issue now, and when I have performance results, I'll post them on this thread if they're good. If they don't meet expectations (which I anticipate as an unlikely outcome at this point), then I'll open a new issue with the performance results instead, referencing this issue.
